Added a WeatherOverlay

TheRaffy opened the pull request:
Closes https://github.com/OpenRA/OpenRA/issues/8687

TheRaffy committed:
Added a WeatherOverlay

pchote commented on OpenRA.Game/Traits/TraitsInterfaces.cs:
```diff
@@ -298,6 +298,7 @@ public interface IRenderInfantrySequenceModifier
 	}
 
 	public interface IPostRender { void RenderAfterWorld(WorldRenderer wr, Actor self); }
+    public interface IRenderWeatherOverlay { void RenderWeatherOverlay(WorldRenderer wr); }
```
Can you use `IPostRender` instead of introducing a new interface for this?

pchote commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,421 @@
+		public readonly string WeatherVariants = "snow";
+		[Desc("Can be true or false")]
+		public readonly bool Wind = true;
+		[Desc("Can be true or false")]
```
The purpose of these descriptions is to describe what they do (what is InstantWind?) - the fact that they can be true or false is obvious from them being bools.

penev92 replied:
Also empty lines between them please.

pchote commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,421 @@
+
+			for (int i = 0; i < count; i++)
+			{
+				if (ChancesPercent(spawnChancePercent))
```
Please just be direct: `if (Game.CosmeticRandom.Next(0, 100) <= spawnChancePercent)` and then remove the unnecessary `ChancesPercent`.

pchote commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,421 @@
+		[Desc("Can be true or false")]
+		public readonly bool WeatherEnabled = true;
+		[Desc("Can be snow or rain or d2kSand")]
+		public readonly string WeatherVariants = "snow";
```
Please make this an enum instead of using magic string values.

pchote commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,421 @@
+		readonly WeatherOverlayInfo info;
+		readonly World world;
+
+		public struct droppingParticle
```
Mutable structs should be avoided in general.

pchote commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,421 @@
+
+		static int CalculateDroppingParticleCount(int x, int y )
+		{  
+			return (x * y) / 1311;  // ~ 800 particles at 1366x768px
```
This parameter probably wants to be exposed via the Info.

penev92 commented on OpenRA.Game/Graphics/WorldRenderer.cs:
```diff
@@ -146,6 +146,10 @@ public void Draw()
 				if (a.Actor.IsInWorld && !a.Actor.Disposed)
 					a.Trait.RenderAfterWorld(this, a.Actor);
 
+            //added for Weathereffects
```
Proper comments start with a space and a capital letter: `// Added for WeatherEffects.`

penev92 commented:
The 3 added maps should probably be unpacked because of the custom rules inside.

penev92 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,421 @@
+			}
+		} 
+		public List<droppingParticle> DroppingParticleList = new List<droppingParticle>();
+		int MaxDrPaCount = CalculateDroppingParticleCount(Game.Renderer.Resolution.Width, Game.Renderer.Resolution.Height);
```
"DrPa" isn't a good naming scheme here or in the other places. In most you can go for "Particles".

penev92 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,421 @@
+			AddADroppingParticle(1, 0, 100);
+
+			// Remove DroppingParticles, which is getting replaced from the top to the bottom, when scrolling down as long as FadeIn
+			// to avoid a gap in the middle of the screen
```
The "as long as FadeIn" part doesn't make sense to me.

TheRaffy replied:
Because I'd like to avoid alot adding and deleting of particles, I'm moveing them to the opposide screen edge ( like in the old Asteroids game).
I fade the count of particles to maxcount by spawning them at y= 0 and they can fall down slowly. The player can now decide to scroll up even the fade in is not complete. Means the particles would get moved from the top to the bottom. and that looks ugly, because of the middle is empty. This part avoids that.

penev92 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,421 @@
+			}
+		}
+
+		droppingParticle XAxisSwing(droppingParticle tempDroppingParticle) {
```
Open blocks on new line please.

penev92 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,421 @@
+					}
+				}
+			}
+			else {  WindType = info.WindType; }
```
Unnecessary `{ }`

abcdefg30 replied:
Also use a new line, please.
```
else
	WindType = info.WindType;

penev92 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,421 @@
+		[Desc("The wind is fading in from zero")]
+		public readonly bool InstantWind = false;
+		[Desc("The wind will change in intensity and direction over time")]
+		public readonly bool AutoWind = true;
```
How about `ChangingWind` or `RandomWind` instead?

penev92 replied:
Also new lines between each property, please.

penev92 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,421 @@
+	public class WeatherOverlayInfo : ITraitInfo
+	{
+		[Desc("On-off switch for the weather overlay")]
+		public readonly bool WeatherEnabled = true;
```
This seems redundant. If people don't want weather they can remove the trait. Also the Info classes don't work as on/off switches.
You can, however, have an `Enabled` property in the trait that you can set during the game, and even expose that to Lua to enable cool stuff to happen.

TheRaffy replied:
It was ment for some lua scripters who like to enable it, after some time has passed.
I don't get the differences of a bool or an Enabled property in that case.

abcdefg30 replied:
If it's meant for lua scripters we don't need to explicitly expose it to yaml.

penev92 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,421 @@
+
+		float TargetWindXOffset = 0f;
+		float CurrentWindXOffset = 0f;
+		public enum WindTypes {
```
You should probably just expose strength and direction in the info instead of having this.

penev92 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,421 @@
+		{
+			if (Game.CosmeticRandom.Next(0, 101) <= percent)
+				return true;
+			else return false;
```
```
else
    return false;
````
please.

abcdefg30 commented on OpenRA.Game/Graphics/WorldRenderer.cs:
```diff
@@ -146,6 +146,10 @@ public void Draw()
 
+            //added for Weathereffects
+            foreach (var item in World.ActorsWithTrait<IRenderWeatherOverlay>())
+                item.Trait.RenderWeatherOverlay(this);         
```
`[SP2000] Invalid spacing at the end of the line.`

TheRaffy commented:
I did just the minor changes, but that's how it will stay until I'll finished the other stuff of my bachelor thesis.

Phrohdoh commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,425 @@
+		public readonly int ParticleDensityFactor = 1311;
+
+		[Desc("Can be snow or rain or d2kSand.")]
+		public readonly string WeatherVariants = "snow";
```
If you have a set limit of what this can be we should provide an Enum and use that type instead of a raw string.

abcdefg30 commented:
Closing as discussed on IRC, @TheRaffy will reopen this once he finds time to fix the remaining issues.

pchote commented:
Added to Future milestone.

TheRaffy commented:
Next Try. I should have fixed all coding reviews. 

(besides the mutable structs, but we talked about that one.)

Unit158 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,527 @@
+		public readonly WeatherOverlay.WeatherVariants  WeatherVariants =WeatherOverlay.WeatherVariants.snow;
+
+		[Desc("Makes dropping particles not falling straight down.")]
+		public readonly bool Wind = true;
```
Can we change this to a vector?

Unit158 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,527 @@
+		[Desc("This works only when ChangingWind is disabled. It sets the wind intensity and direction (sign) manually. Values between -5 and 5 are used when ChangingWind is enabled.")]
+		public readonly int WindStrength;
+		
+		[Desc("The first color of the snow particles. The default is the 128th. Color from the temperate.pal")]
```
Can we get an array of Color instead? That way we will only have one color field.

Unit158 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,527 @@
+		// Adds Particles scattered over the Y-Axis range with a spawnchance of %
+		void AddAParticle(int count, int rangeY, int spawnChancePercent)
+		{
+
```
Invalid newline

Unit158 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,527 @@
+				{
+					if (info.WeatherVariants == WeatherVariants.snow)
+					{
+						var tempColor= Color.Pink;
```
Needs a space.

Unit158 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,527 @@
+	[Desc("Adds an overlay for weather animations based on dropping particles like snowfalkes, raindrops or sand grains.")]
+	public class WeatherOverlayInfo : ITraitInfo
+	{
+		[Desc("Factor for particle density. As smaller the value as more partikles will get spawned.")]
```
"A smaller value increases the number of particles spawned."

Unit158 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,527 @@
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Adds an overlay for weather animations based on dropping particles like snowfalkes, raindrops or sand grains.")]
```
"Adds a particle-based overlay."

We don't really want hardcoded particle types.

Unit158 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,527 @@
+		public readonly bool ChangingWind = true;
+
+		[Desc("This works only when ChangingWind is disabled. It sets the wind intensity and direction (sign) manually. Values between -5 and 5 are used when ChangingWind is enabled.")]
+		public readonly int WindStrength;
```
If the above is changed to a vector (or a series of vectors for a swaying effect?) this will no longer be needed.

Unit158 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,527 @@
+		readonly WeatherOverlayInfo info;
+		readonly World world;
+		public enum WeatherVariants { snow, rain, d2kSand };
+		public struct Particle
```
This class has a huge amount of state. Is it possible to reduce the amount of variables you use?

TheRaffy replied:
any ideas?

Unit158 replied:
@chrisforbes would actually have a better idea than I on how to reduce this.

Unit158 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,527 @@
+		{
+			public float PosX;
+			public float PosY;
+			public int   Size;
```
Identifiers need not be aligned.

Unit158 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,527 @@
+			public float Gravity;
+			public float SwingOffset;
+			public float SwingSpeed;
+			public bool  SwingDirection;
```
Here too.

Unit158 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,527 @@
+		}
+
+		// Adds Particles scattered over the Y-Axis range with a spawnchance of %
+		void AddAParticle(int count, int rangeY, int spawnChancePercent)
```
AddAParticle -> SpawnParticles

Unit158 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,527 @@
+			return (x * y) / tempfactor;  
+		}
+
+		// Adds Particles scattered over the Y-Axis range with a spawnchance of %
```
This comment isn't needed.

Unit158 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,527 @@
+
+				ParticleList[i] = tempParticle;
+			}
+			anitScrollPrevTopLeft = wr.Viewport.TopLeft;
```
anit?

Unit158 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,527 @@
+			return tempParticle;
+		}
+
+		// AnitScroll keeps the DroppingPartikels in place when scrolling the viewport
```
Partikels -> Particles

Unit158 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,527 @@
+		}
+
+		// AnitScroll keeps the DroppingPartikels in place when scrolling the viewport
+		Particle AnitScroll(Particle tempParticle, WorldRenderer wr)
```
AntiScroll

Unit158 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,527 @@
+		public readonly bool Wind = true;
+
+		[Desc("The wind is fading in from zero.")]
+		public readonly bool InstantWind = false;
```
This should probably be a user-configurable delay.

Unit158 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,527 @@
+		public bool Enabled = true;
+		readonly WeatherOverlayInfo info;
+		readonly World world;
+		public enum WeatherVariants { snow, rain, d2kSand };
```
This should be replaced by fully user-configurable particles.

Unit158 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,527 @@
+
+	public class WeatherOverlay : ITick, IPostRender
+	{
+		public bool Enabled = true;
```
No code can disable this, so this should be hooked up to an upgrade.

Phrohdoh replied:
Why do you say no code can disable this?

Unit158 replied:
I meant that nothing disables this at the moment (aside from the info?)

TheRaffy replied:
Unit was right. It was ment for some lua scripters, but I already removted it.

Phrohdoh replied:
I did make use of this in Lua scripting, but haven't brought this up because it hasn't been merged yet.

Unit158 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,527 @@
+		};
+		public WindTypes WindType = WindTypes.Calm;
+		long windTickCountdown = 1500;
+		float2 anitScrollPrevTopLeft;
```
anit -> anti

TheRaffy commented:
Hey, new update here. Hope its all good now.
Will not be able to fix any big stuff for november.
greetings

abcdefg30 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+		[Desc("The wind is fading softly from zero or instant.")]
+		public readonly bool InstantWindChanges = false;
+
+		[Desc("Particles are drawn with rectangles when enable, else with lines.")]
```
`enable` -> `enabled`

abcdefg30 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+		public readonly bool InstantWindChanges = false;
+
+		[Desc("Particles are drawn with rectangles when enable, else with lines.")]
+		public readonly bool RectOrLine = true;
```
I also don't like the naming of this bool.
`true` implies that this is using `Rect`s and `Line`s (not only `Rect`s).
`false` that this is using neither.

How about something like `UseRectangles`? The description gives then away what happens when this is `false`.

abcdefg30 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+		[Desc("Works only with rectangles enabled. Size min. and max. value.")]
+		public readonly int[] PaticleSize = { 1, 3 };
+
+		[Desc("Scatters falling direction on the x-axis . Scatter min. and max. value.")]
```
Trailing space after `x-axis`.

abcdefg30 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+
+		[Desc("The colors for the particles.",
+			"For snow and rain use the temperate.pal with an index of 128-131 and 164-167.",
+			"and for a sandstorm at Dune2k the d2k.pal with an index of 78, 82, 85, 89.")]
```
I'm not sure if we want mod specific content in the `Desc`s.
We could document that e.g. on a wiki page or just give a map as example (as you actually already do).

abcdefg30 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+			var tempfactor = 0;
+			if (info.ParticleDensityFactor != 0)
+				tempfactor = info.ParticleDensityFactor;
+			else tempfactor = 1;
```
`tempfactor = 1;` on a new line, please.

abcdefg30 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+				tempfactor = info.ParticleDensityFactor;
+			else tempfactor = 1;
+
+			return (x * y) / tempfactor;
```
You could actually simplify the method:
`return (x * y) / (info.ParticleDensityFactor != 0 ? info.ParticleDensityFactor : 1)`

abcdefg30 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+				{
+					var tempColor = info.ParticleColors[Game.CosmeticRandom.Next(0, info.ParticleColors.Length / 4 + 1)];
+					var tempColorTail = Color.FromArgb(info.LineTailAlphaValue, tempColor.R, tempColor.G, tempColor.B);
+					var tempSwingDirection = Game.CosmeticRandom.Next(0, 2) == 0 ? true : false;
```
`var tempSwingDirection = Game.CosmeticRandom.Next(0, 2) == 0;` should be sufficient.

abcdefg30 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+
+		Particle WindLogic(Particle tempParticle)
+		{
+			if (info.ChangingWind)
```
Could we get `if (info.ChangingWind && windTickCountdown <= 0)` here and remove the check below to reduce the nestling?

abcdefg30 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+						currentWindXOffset += 0.00001f;
+				}
+			}
+			else { currentWindXOffset = targetWindXOffset; }
```
`{ currentWindXOffset = targetWindXOffset; }` on a new line, please.

abcdefg30 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+
+			// Y-Axis up
+			if (tempParticle.PosY < 0)
+
```
We need braces here.

abcdefg30 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+			return tempParticle;
+		}
+
+		// Update
```
Unnecessary comment (the method name contains `Update`).

abcdefg30 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+
+		public void RenderAfterWorld(WorldRenderer wr, Actor self)
+		{
+			// Update
```
Same here.

abcdefg30 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+				var tempPos = new float2(item.PosX + topLeft.X, item.PosY + topLeft.Y);
+
+				if (info.RectOrLine)
+				Game.Renderer.WorldQuadRenderer.FillRect(new RectangleF(tempPos.X, tempPos.Y, item.Size, item.Size), item.Color);
```
Please add one level of indentation here.

abcdefg30 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+		public readonly int[] PaticleSize = { 1, 3 };
+
+		[Desc("Scatters falling direction on the x-axis . Scatter min. and max. value.")]
+		public readonly int[] DirectionScatter = { -1, 1 };
```
How about `ScatterDirection`?

abcdefg30 commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+		Particle XAxisSwing(Particle tempParticle)
+		{
+			// Left direction turn
+			if (tempParticle.SwingOffset < (-1 * tempParticle.SwingAmplitude))
```
`-1 * tempParticle.SwingAmplitude` -> `-tempParticle.SwingAmplitude`.

pchote commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+			public Color Color;
+			public Color TailColor;
+
+			public Particle(
```
You may be able to remove the constructor completely and then use initializer syntax below:
```csharp
new Particle
{
    PosX = Game.CosmeticRandom.Next(0, Game.Renderer.Resolution.Width + 1),
    PosY = Game.CosmeticRandom.Next(0, rangeY + 1),
```
and so on.

TheRaffy replied:
I failed to do this one.

pchote commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+
+		void SpawnParticles(int count, int rangeY, int spawnChancePercent)
+		{
+			for (int i = 0; i < count; i++)
```
`var`

pchote commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+		{
+			for (int i = 0; i < count; i++)
+			{
+				if (Game.CosmeticRandom.Next(0, 101) <= spawnChancePercent)
```
`if (Game.CosmeticRandom.Next(100) < spawnChancePercent)`

pchote commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+			{
+				if (Game.CosmeticRandom.Next(0, 101) <= spawnChancePercent)
+				{
+					var tempColor = info.ParticleColors[Game.CosmeticRandom.Next(0, info.ParticleColors.Length / 4 + 1)];
```
You should remove the initial `0` from all these random.Next calls.

TheRaffy replied:
That's clever.

pchote commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+							info.DirectionScatter[0] + Game.CosmeticRandom.Next(0, info.DirectionScatter[1] - info.DirectionScatter[0] + 1),
+							(info.Gravity[0] + Game.CosmeticRandom.Next(0, info.Gravity[1] - info.Gravity[0] + 1)) / 100f,
+							Game.CosmeticRandom.Next(info.SwingOffset[0], info.SwingOffset[1] + 1) / 10f,
+							Game.CosmeticRandom.Next(info.SwingSpeed[0], info.SwingSpeed[1] + 1) / 1000f,
```
What are these constants of 10, 1000 doing?

TheRaffy replied:
That Game.CosmeticRandom.Next() does not support float. -> thats how i get my floats

pchote replied:
What do the magnitudes of the divisors represent?

pchote commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+			// Remove Particles, which are getting replaced from the top to the bottom by the "EdgeCheckReplace",
+			// when scrolling down, as long as the FadeIn is not completed,
+			// to avoid having particles at the top and bottom, but not in the middle of the screen.
+			for (int i = 0; i < ParticleList.Count; i++)
```
`var`

pchote commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+
+		void FadeOutCountDoPa(WorldRenderer wr)
+		{
+			for (int i = 0; i < ParticleList.Count; i++)
```
`var`

pchote commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+
+		Particle Movement(Particle tempParticle)
+		{
+			tempParticle.PosX = tempParticle.PosX + tempParticle.DirectionScatterX + tempParticle.SwingOffset + currentWindXOffset;
```
`+=`

pchote commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+		// AntiScroll keeps the particels in place when scrolling the viewport
+		Particle AntiScroll(Particle tempParticle, WorldRenderer wr)
+		{
+			if (antiScrollPrevTopLeft != wr.Viewport.TopLeft) {
```
Opening braces on their own line, please.  We don't use egyptian braces here :wink:

pchote commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+		{
+			ParticlesCountLogic(wr);
+
+			for (int i = 0; i < ParticleList.Count; i++)
```
`var`

pchote commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+				UpdateWeatherOverlay(wr);
+
+			// Draw
+			float2 topLeft = wr.Viewport.TopLeft;
```
`var`

pchote commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+
+			// Draw
+			float2 topLeft = wr.Viewport.TopLeft;
+			foreach (Particle item in ParticleList)
```
`var`

pchote commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+		public readonly int[] DirectionScatter = { -1, 1 };
+
+		[Desc("The value that particles fall down. Gravity min. and max. value.")]
+		public readonly int[] Gravity = { 100, 200 };
```
What are the units of all these constants? (related to my question below about constants - I suspect these are arbitrary).

TheRaffy replied:
Gravity will get divided by 100 later, its all about the comma.
Game.CosmeticRandom.Next() does not support floats.
I also could use floats here,  multiply them later, do the random and divide them again.
But would be more to calculate.

pchote commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+					if (antiScrollPrevTopLeft.Y > wr.Viewport.TopLeft.Y)
+						tempParticle.PosY += antiScrollPrevTopLeft.Y - wr.Viewport.TopLeft.Y;
+					else
+						tempParticle.PosY -= wr.Viewport.TopLeft.Y - antiScrollPrevTopLeft.Y;
```
This inner if/else is redundant.  The same evaluation is done in both cases.

pchote commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+					if (antiScrollPrevTopLeft.X > wr.Viewport.TopLeft.X)
+						tempParticle.PosX += antiScrollPrevTopLeft.X - wr.Viewport.TopLeft.X;
+					else
+						tempParticle.PosX -= wr.Viewport.TopLeft.X - antiScrollPrevTopLeft.X;
```
This inner if/else is redundant.  The same evaluation is done in both cases.

pchote commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+			public float Gravity;
+			public float SwingOffset;
+			public float SwingSpeed;
+			public bool SwingDirection;
```
A `bool` isn't a direction.  Please rename this to e.g. SwingRight.

pchote replied:
A `bool` isn't a direction.  Please change this to `public int SwingDirection` (assigned values of + or - 1) and then use it as a multiplier instead of having separate if/else cases everywhere.

pchote commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+
+			// Right direction turn
+			else if (tempParticle.SwingOffset > tempParticle.SwingAmplitude)
+				tempParticle.SwingDirection = !tempParticle.SwingDirection;
```
`tempParticle.SwingDirection ^= true`

pchote commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+			}
+		}
+
+		void FadeOutCountDoPa(WorldRenderer wr)
```
Please use a more descriptive method name.

pchote commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+		{
+			// X-Axis right
+			if (tempParticle.PosX > Game.Renderer.Resolution.Width)
+				tempParticle.PosX -= Game.Renderer.Resolution.Width;
```
You should be able to replace this and the if below it with `tempParticle.PosX %= Game.Renderer.Resolution.Width`.  You can then do the same for height.

pchote commented on OpenRA.Game/Graphics/WorldRenderer.cs:
```diff
@@ -141,7 +141,7 @@ public void Draw()
 				renderables[i].Render(this);
 
-			// added for contrails
+			// Added for contrails and WeatherOverlay
```
Please just remove this comment.  The engine code shouldn't be referring to traits defined in the mod code.

WolfGaming commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+					}
+					else
+					{
+						if (currentWindIndex < info.WindStrength.Length - 1)
```
Please change the above to an else if.

WolfGaming commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+		{
+			if (info.ChangingWind)
+			{
+				if (windTickCountdown <= 0)
```
if you change the if to !info.ChangingWind and switch the if else statement you can change the above to else if (windTickCountdown <= 0).

WolfGaming commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+			// Fading the wind in little steps towards the TargetWindOffset
+			if (!info.InstantWindChanges)
+			{
+				if (currentWindXOffset != targetWindXOffset)
```
same here, change !info to info, switch the if and else body and change the else to else if (currentWindXOffset != targetWindXOffset).

WolfGaming commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+		Particle Movement(Particle tempParticle)
+		{
+			tempParticle.PosX = tempParticle.PosX + tempParticle.DirectionScatterX + tempParticle.SwingOffset + currentWindXOffset;
+			tempParticle.PosY = tempParticle.PosY + tempParticle.Gravity;
```
+= here too.

WolfGaming commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+				var tempParticleCount = CalculateParticleCount(Game.Renderer.Resolution.Width, (int)tempRangeY);
+				if (ParticleList.Count + tempParticleCount > maxParticleCount)
+					tempParticleCount = maxParticleCount - ParticleList.Count;
+				SpawnParticles(tempParticleCount, (int)tempRangeY, 50);
```
new line after the if body.

WolfGaming commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+			// Left direction turn
+			if (tempParticle.SwingOffset < (-1 * tempParticle.SwingAmplitude))
+				tempParticle.SwingDirection = !tempParticle.SwingDirection;
+
```
random new line?

TheRaffy commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,391 @@
+		[Desc("The colors for the particles.",
+			"For snow and rain use the temperate.pal with an index of 128-131 and 164-167.",
+			"and for a sandstorm at Dune2k the d2k.pal with an index of 78, 82, 85, 89.")]
+		public readonly Color[] ParticleColors = {
```
Did a bug here That caused all particles to use the default values, I renamed it here but not in the yaml..

TheRaffy commented:
New update! Thanks to abcdef and Pchote
+ fixed a bug with the colors

TheRaffy commented:
there appeared new comments. have to update again

TheRaffy commented:
New Update here.
Thanks a lot for the reviews.
I was able to implement everything suggested.

pchote commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,327 @@
+			for (var i = 0; i < count; i++)
+			{
+				if (Game.CosmeticRandom.Next(100) <= spawnChancePercent)
+				{
```
This needs to be < (not <=) to give the expected behaviour.

pchote commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,327 @@
+			{
+				if (Game.CosmeticRandom.Next(100) <= spawnChancePercent)
+				{
+					var tempColor = info.ParticleColors[Game.CosmeticRandom.Next(info.ParticleColors.Length)];
```
We have specific functions for picking an element in a collection,
IIRC it is something like info.ParticleColors.Random(Game.CosmeticRandom)

pchote commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,327 @@
+		void FadeOutParticleCount(WorldRenderer wr)
+		{
+			for (var i = 0; i < ParticleList.Count; i++)
+			{
```
These braces can be omitted.

pchote commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,327 @@
+		{
+			// Direction turn
+			if (tempParticle.SwingOffset < -tempParticle.SwingAmplitude || tempParticle.SwingOffset > tempParticle.SwingAmplitude)
+				tempParticle.SwingDirection = -tempParticle.SwingDirection;
```
tempParticle.SwingDirection *= -1

pchote commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,327 @@
+			if (tempParticle.SwingOffset < -tempParticle.SwingAmplitude || tempParticle.SwingOffset > tempParticle.SwingAmplitude)
+				tempParticle.SwingDirection = -tempParticle.SwingDirection;
+
+			// Performe the X-Axis-Swing
```
Typo in comment

pchote commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,327 @@
+			else if (windTickCountdown <= 0)
+			{
+				windTickCountdown = Game.CosmeticRandom.Next(info.WindTick[0], info.WindTick[1]);
+				if (Game.CosmeticRandom.Next(1) == 1 && currentWindIndex > 0)
```
.Next returns at most 1 less than the given argument, so this will always be false.

pchote commented on mods/ra/maps/singles-in-the-rain/map.yaml:
```diff
@@ -0,0 +1,170 @@
+MapFormat: 7
```
Fort Lonestar would be a better test case than duplicating an existing map.

TheRaffy replied:
I'd like to see how the weather performs with scrolling at a bigger map. The F.Lonestar funmap is to small for nice radar based viewport jumps. Theres also no ocean at F. Lonestar. And I'd also like to see the droplets over the watersurface and future more maybe there will be these littel raindrop circles on the water surface in the future (Not sure yet). And Fort Lonestar is using a night setting, what would cause me to use darker particle colors.
That's why I think F.Lonestar is not the better map.

pchote replied:
That's fine for testing, but once this has been demonstrated there's no need to include them.  I think it is a bad idea to ship duplicated maps in our main distribution.

Fort lonestar is set during a thunderstorm and is already configured with rain sound effects.  It will be weird to not have this feature used in it.

pchote commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,327 @@
+		public readonly bool ChangingWind = true;
+
+		[Desc("Works only if ChangingWind is enabled. The values of the wind intensity and direction. ")]
+		public readonly int[] WindStrength = { -5, -3, -2, 0, 2, 3, 5 };
```
Following in from my earlier questions/comments, please use an intuitive unit for this (such as pixels per tick - float values are fine since these are visual) instead of arbitrary units with an arbitrary divisor in the code.

pchote replied:
And likewise for the other parameters.

TheRaffy commented:
New Update.

Mailaender commented:
The results look spectacular. Can you move the test cases onto existing maps (e.g. Fort Lonestar, Snow Town) and fix the two remaining style comments?

RoosterDragon commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,325 @@
+		[Desc("Particles are drawn with rectangles when enabled, else with lines.")]
+		public readonly bool UseRectangles = true;
+
+		[Desc("Works only with rectangles enabled. Size  min. and max. value in pixels.")]
```
Errant double space after Size.

RoosterDragon commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,328 @@
+			return tempParticle;
+		}
+
+		// AntiScroll keeps the particels in place when scrolling the viewport
```
particels -> particles

RoosterDragon commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,328 @@
+				tempParticle = WindLogic(tempParticle);
+				tempParticle = Movement(tempParticle);
+				tempParticle = AntiScroll(tempParticle, wr);
+				tempParticle = EdgeCheckReplace(tempParticle, wr);
```
Consider making all the methods return void and pass the particle by reference. This avoids having to copy the (rather large) struct. I found this change made ```UpdateWeatherOverlay``` twice as fast.

There seem to be some comments earlier regarding the mutability of the struct, so I'm assuming it was decided we needed a mutable struct and not a class.

RoosterDragon commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,328 @@
+			public Color TailColor;
+		}
+
+		public List<Particle> ParticleList = new List<Particle>();
```
This doesn't need to be public.

RoosterDragon commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,328 @@
+	{
+		readonly WeatherOverlayInfo info;
+		readonly World world;
+		public struct Particle
```
This doesn't need to be public.

RoosterDragon commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,328 @@
+
+		void ParticlesCountLogic(WorldRenderer wr)
+		{
+			// Logic
```
This comment doesn't tell me anything.

RoosterDragon commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,328 @@
+					tempParticle.PosX += antiScrollPrevTopLeft.X - wr.Viewport.TopLeft.X;
+
+				if (antiScrollPrevTopLeft.Y != wr.Viewport.TopLeft.Y)
+					tempParticle.PosY += antiScrollPrevTopLeft.Y - wr.Viewport.TopLeft.Y;
```
It doesn't seem we need any of these three if statements here - the positions would just get updated by zero.

I'd suggest removing them for clarity.

TheRaffy replied:
That would be faster, too?

TheRaffy commented:
Fixed the two things that Mailänder mentioned and that stuff what rooster said. 
- Removed the singles in the rain and doubles in the snow and the sandstorm testmap
- Snow town and Fortlonestar are using now the weather particles

Thanks!

Phrohdoh commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,314 @@
+	public class WeatherOverlayInfo : ITraitInfo
+	{
+		[Desc("Factor for particle density. A smaller value increases the number of particles spawned.")]
+		public readonly int ParticleDensityFactor = 1311;
```
The logic here is inverted. If a density value is high then a large amount of particles ought to exist, currently this is the other way around.

penev92 replied:
Then perhaps `ParticleSpreadFactor` ?
Or perhaps keep the clearer name and invert the logic that uses it.

Phrohdoh commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,314 @@
+		[Desc("Factor for particle density. A smaller value increases the number of particles spawned.")]
+		public readonly int ParticleDensityFactor = 1311;
+
+		[Desc("The wind will change over time by the values of WindStrength, or just stick to the first value.")]
```
Please reword this to `Should the force of the wind change over time?` (or replace `force` with `power`).

Phrohdoh commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,314 @@
+		[Desc("The wind will change over time by the values of WindStrength, or just stick to the first value.")]
+		public readonly bool ChangingWind = true;
+
+		[Desc("Works only if ChangingWind is enabled. The levels of wind intensity desc. the x-movement value of the particles in px/tick.")]
```
I don't know why ` desc.` is in this description or what it represent.

GraionDilach replied:
"describes" - although this probably should have a line break somewhere.

Phrohdoh commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,314 @@
+		[Desc("Works only if ChangingWind is enabled. The levels of wind intensity desc. the x-movement value of the particles in px/tick.")]
+		public readonly int[] WindStrength = { -5, -3, -2, 0, 2, 3, 5 };
+
+		[Desc("Works only if ChangingWind is enabled. Min. and max. ticks needed to change the wind. ")]
```
I don't understand what these values are for and this description isn't helping.
I'll try to come up with a suggestion.

TheRaffy replied:
The wind intensity is changing after a duration, this values are the min and max value of this time.
( the Index of the WindStrength  will eighter increase or decrease.)

Phrohdoh commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,314 @@
+		[Desc("Works only if ChangingWind is enabled. Min. and max. ticks needed to change the wind. ")]
+		public readonly int[] WindTick = { 150, 750 };
+
+		[Desc("The wind is fading softly from zero or instant.")]
```
I don't understand what this changes, but the description seems to conflict with `ChangingWind`.

If I set `ChangingWind: false` I expect the wind force to never change.

Phrohdoh commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,314 @@
+		[Desc("Particles are drawn with rectangles when enabled, else with lines.")]
+		public readonly bool UseRectangles = true;
+
+		[Desc("Works only with rectangles enabled. Size min. and max. value in pixels.")]
```
Size as in each value is width and height?
If so, it isn't obvious because you use rectangles not squares.

Phrohdoh commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,314 @@
+		[Desc("Scatters falling direction on the x-axis. Scatter min. and max. value in px/tick.")]
+		public readonly int[] ScatterDirection = { -1, 1 };
+
+		[Desc("The value that particles fall down. Gravity min. and max. value in px/tick.")]
```
`Min and Max speed at which particles fall in px/tick.`

Phrohdoh commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,314 @@
+		[Desc("The value range that can be swung to the left or right. SwingAmplitude min. and max. value in px/tick.")]
+		public readonly float[] SwingAmplitude = { 1.0f, 1.5f };
+
+		[Desc("The ArgbColors for the particles. Think about using the colors of the temperate.pal for example.")]
```
`Color` is known to be ARGB.
This doesn't tell me how each item in the array is used so I wouldn't know proper order, etc.

Unit158 replied:
Color is usually HSL in OpenRA.

Phrohdoh commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,314 @@
+			Color.FromArgb(255, 188, 188, 188)
+		};
+
+		[Desc("Works only with line enabled and will fade out the tail of the line.")]
```
Isn't this pretty much a contrail?
If so I think that's worth mentioning as it is an easy concept to follow.

TheRaffy replied:
Contrails can have curves. Weather particles have not. both are lines, true. But besides that its like mentioning the fog of war because it's also transparent. So i guess the word "transparent" is missing in the Desc.

Phrohdoh commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,314 @@
+		}
+
+		List<Particle> particleList = new List<Particle>();
+		int maxParticleCount;
```
Can this be `readonly`?

TheRaffy replied:
This value get created by the ParticleDensityFactorand and theScreensizes. 
To get it randomed would not make much sence at the current state in my opinion.
If we like to play with the particle count, it would make more sence to get a min and a max count and fade between them over a periode of time while the match is running. Like I already do it with the wind. That is something I'm aware of.

Phrohdoh replied:
I'm not asking for random.
[MSDN readonly](https://msdn.microsoft.com/en-us/library/acdd6hb7.aspx)

TheRaffy replied:
lol.. sry, I , truely read random for some reason.

Phrohdoh commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,314 @@
+		{
+			for (var i = 0; i < count; i++)
+			{
+				if (Game.CosmeticRandom.Next(100) < spawnChancePercent)
```
Please reverse the order of these two values.

Phrohdoh commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,314 @@
+			// when scrolling down, as long as the FadeIn is not completed,
+			// to avoid having particles at the top and bottom, but not in the middle of the screen.
+			for (var i = 0; i < particleList.Count; i++)
+			{
```
Unnecessary brackets.

Phrohdoh commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,314 @@
+				if (particleList.Count + tempParticleCount > maxParticleCount)
+					tempParticleCount = maxParticleCount - particleList.Count;
+
+				SpawnParticles(tempParticleCount, (int)tempRangeY, 50);
```
Where is this `50` coming from? Can it be exposed to yaml?

TheRaffy replied:
its a 50:50 spawn chance. Other Values would not be very usefull and like to look ugly. 
It's like scrolling upwards and getting a brick of white snowflakes in the face, or erasing lines in the  same way.

Phrohdoh commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,314 @@
+
+		void EdgeCheckReplace(ref Particle tempParticle, WorldRenderer wr)
+		{
+			// X-Axis
```
These comments are unnecessary.

Phrohdoh commented:
I haven't asked for any real logic changes so :+1: from me.
This feature rocks and looks great in-game!
Good job @TheRaffy.

I'd like to see my inline comments at least addressed.

![screen shot 2015-08-26 at 8 53 11 pm](https://cloud.githubusercontent.com/assets/4861023/9510703/9ec3b594-4c34-11e5-9826-42de8fd08ad3.png)

TheRaffy commented:
Next update. 
Changed the stuff what you recommended phrohdoh.
I also renamed the wind parameters, hope it's better now.
But I didn't expand the WeatherOverlay to a dynamic particle count.

Phrohdoh commented on OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs:
```diff
@@ -0,0 +1,310 @@
+		[Desc("The value range that can be swung to the left or right. SwingAmplitude min. and max. value in px/tick.")]
+		public readonly float[] SwingAmplitude = { 1.0f, 1.5f };
+
+		[Desc("The randomly selected ArgbColors for the particles. Use this order: a,r,g,b,  a,r,g,b, a,...")]
```
This isn't quite what I meant. `ARGB` already tells the order.
Modders don't know what the colors are for. Moreso why do you want an array rather than a single value?

TheRaffy replied:
Different shades of color creates an illusion of depth. Its like you ask why the ocean is using more than 1 color. Well its looks better I guess. I thought about an automatic shading, but this here is more powerfull, because you can also use different colors. And future more I need to use the exact colors of the palett, the white colors of the palett is not faded in an linar way. Like always -16 to r,g,b.

TheRaffy commented:
fixed the readonly, didn't fix the color desc.

Mailaender commented:
Needs a rebase.

TheRaffy commented:
rebased

Unit158 commented:
Looks very pretty. :+1:

RoosterDragon commented:
Can you remove changes to ```OpenRA.sln``` from the commit please?

TheRaffy commented:
removed OpenRA.sln from the commit
renamed commit to: -Added a WeatherOverlay

TheRaffy commented:
Renamed commit to: Added a WeatherOverlay
Anything else maybe serving some crinkle-cut French fries?

RoosterDragon commented:
```OpenRA.sln``` has snuck back in (VS will automatically update this all the time, so you to keep an eye on it).

TheRaffy commented:
removed OpenRA.sln from the commit again.

RoosterDragon commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/5481f9aca23f0d7e23d2134ef3b820d62f664393...5a179fa592d7d1ba382ec291bb28685aa9c87f10)

GrayHatter commented:
Did this cause the performance bug?

TheRaffy commented:
There is a performance graph which you can enable to gain more informations.
On what map did you play? and what "performance bug" are you talking about?

GrayHatter commented:
Sorry, posted without this #9839. (The issue that I was referring to)

I talked to @JFreegman, he's the one with this issue, and he thinks it's very unlikely, that these two are related.

RoosterDragon commented:
This isn't the cause of #9839 - only maps that make use of the weather feature pay the performance cost (which isn't super-bad or anything anyway).
